### PR TITLE
add markdown widget

### PR DESCRIPTION
**Pull request by elevenhsoft**
Hello,

This PR extends text module with new markdown widget. Everything is hidden behind `markdown` feature which needs to be enabled to work with this widget. Access is from `widget::text::markdown()` via helper function. This function require passing text in markdown format as a `String`. The markdown `parser` and `Markdown` widget are in different files because it's easier to work with it. From my testing it works pretty good, the text is formatted. There are some problems with links and images, so I made them as a text with URL (not clickable ofc), the same with image.

The second problem and I don't know how to solve it, is with strings which comes from `markdown` crate. For spans in `rich_text()` we need string slices (or no?) so I used `Box::leak()` to achieve this. Maybe someone have better idea, or ideal would be if spans can contains `Strings`. But I did not test it yet.

### CODE REVIEW

File: examples/markdown-widget/Cargo.toml
Comment (mmstick) [inline examples/markdown-widget/Cargo.toml:1]: This example should be called `markdown`

**Comment by wiiznokes** [comment]
Maybe this could be added directly in Iced, this is on the Iced roadmap: https://whimsical.com/roadmap-iced-7vhq6R35Lp3TmYH4WeYwLM

File: examples/markdown-widget/Cargo.toml
Comment (elevenhsoft) [reply examples/markdown-widget/Cargo.toml:1]: Well, that's what I did before this rename. However I don't know what's happened but it did not work with `markdown` crate, there was a dependency issue. I'm not sure but renaming this example fixed this issue :sweat_smile:

**Comment by mmstick** [comment]
I'll check the string problem

**Comment by elevenhsoft** [comment]
> Maybe this could be added directly in Iced, this is on the Iced roadmap: https://whimsical.com/roadmap-iced-7vhq6R35Lp3TmYH4WeYwLM

Yeah, I'm here after this comment: https://github.com/elevenhsoft/cosmic-ext-applet-ollama/issues/1#issuecomment-2168355798 ^^

I did know about Iced roadmap. They need to update cosmic-text tho if I'm correct.

**Comment by wiiznokes** [comment]
Putting it in Iced would be better, but it will probably be merged after 2 years :smile:, so nevermind my coment. I may use it in fan-control for an explanation screen so its nice to see this :)

Btw, are you using fedora ? I tried to package your applet 2 hours ago but i can't install Ollama.

**Comment by elevenhsoft** [comment]
> Btw, are you using fedora ? I tried to package your applet 2 hours ago but i can't install Ollama.

No, only Pop_OS and openSUSE.

**Comment by elevenhsoft** [comment]
Hey @mmstick , I made some changes to Parser, tried to remove most of the `Box::leak()` with success. Some of them are still there but should not cause issues I hope :)

**Comment by mmstick** [comment]
I will try to eliminate the remaining leaks soon.

**Review by jackpot51 — CHANGES_REQUESTED** [review]
This is a lot of code, and the code from cosmic-edit for rendering a Buffer is not ideal. I want to help optimize this before merging to ensure it renders using iced's Buffer rendering code so it can be GPU accelerated.

**Comment by elevenhsoft** [comment]
Thank you! :)

**Comment by wiiznokes** [comment]
In case anyone missed it, there is now a markdown widget in iced: https://github.com/iced-rs/iced/pull/2508

**Comment by elevenhsoft** [comment]
> In case anyone missed it, there is now a markdown widget in iced: [iced-rs/iced#2508](https://github.com/iced-rs/iced/pull/2508)

Thank you!

In this case I think we should use Iced's markdown. This one is... well, far from being good. There is still a lot of work to be done.

**Pull request closed without merging**